File: PO_py/program/Panel.py
import tkinter as tk

from program.Organisms import Animal
from program.Organisms.Animals import Sheep


class Panel(tk.Canvas):
    #def __init__(self, master, height, width, rect_size, *args, **kwargs):
    def __init__(self, master, height, width, rect_size, world, *args, **kwargs):
        super().__init__(master, *args, **kwargs)
        self.width = width
        self.height = height
        self.rect_size = rect_size
        self.world = world
        self.config(width=width * rect_size, height=(10 + height) * rect_size + 10)
        self.draw_world()

    def draw_world(self):
        self.delete("all")
        """
        for i in range(self.height):
            for j in range(self.width):
                organism = self.world.get_organism_at_position(j, i)
                if isinstance(organism, Animal):
                    self.create_oval(
                        j * self.rect_size, i * self.rect_size,
                        (j + 1) * self.rect_size, (i + 1) * self.rect_size,
                        fill=self.get_color_for_organism(organism),
                        outline="black"
                    )

                if isinstance(organism, Plant):
                    self.create_rectangle(
                        j * self.rect_size, i * self.rect_size,
                        (j + 1) * self.rect_size, (i + 1) * self.rect_size,
                        fill="green"
                    )
                    color = self.get_color_for_organism(organism)
                    for k in range(4):
                        dot_x = j * self.rect_size + self.rect_size / 4 + (k % 2) * self.rect_size / 2 - self.rect_size / 8
                        dot_y = i * self.rect_size + self.rect_size / 4 + (k // 2) * self.rect_size / 2 - self.rect_size / 8
                        if isinstance(organism, (DeadlyNightshade, Guarana, Barszcz)):
                            self.create_oval(
                                dot_x, dot_y,
                                dot_x + self.rect_size / 6, dot_y + self.rect_size / 6,
                                fill=color
                            )
                        if isinstance(organism, Dandelion):
                            self.create_line(dot_x, dot_y, dot_x + self.rect_size / 6, dot_y + self.rect_size / 6)
                            self.create_line(dot_x + self.rect_size / 6, dot_y, dot_x, dot_y + self.rect_size / 6)
                            self.create_line(dot_x + self.rect_size / 12, dot_y, dot_x + self.rect_size / 12, dot_y + self.rect_size / 6)
                            self.create_line(dot_x, dot_y + self.rect_size / 12, dot_x + self.rect_size / 6, dot_y + self.rect_size / 12)
    """
        self.create_rectangle(10, 10, self.rect_size * self.width, self.rect_size * self.height, outline="black")
        self.create_rectangle(10, self.rect_size * (self.width + 1), self.rect_size * self.width, self.rect_size * 3, outline="black")
        message = f"Aktualna tura: {self.world.get_turn()}"
        self.create_text(self.rect_size, self.rect_size * (self.width + 2), text=message, anchor=tk.NW)


    def refresh(self):
        self.draw_world()


    def get_color_for_organism(self, organism):
        #if isinstance(organism, Wolf):
            #return "gray"
        if isinstance(organism, Sheep):
            return "white"
"""        elif isinstance(organism, Turtle):
            return "darkgreen"
        elif isinstance(organism, Fox):
            return "orange"
        elif isinstance(organism, Antelope):
            return "yellow"
        elif isinstance(organism, Human):
            return "pink"
        elif isinstance(organism, Guarana):
            return "red"
        elif isinstance(organism, Dandelion):
            return "yellow"
        elif isinstance(organism, DeadlyNightshade):
            return "magenta"
        elif isinstance(organism, Barszcz):
            return "white"
        return "black"
"""

File: PO_java/program/Panel.java
package projekt.program;
import projekt.program.Organisms.Animal;
import projekt.program.Organisms.Animals.*;
import projekt.program.Organisms.Plant;
import projekt.program.Organisms.Plants.Barszcz;
import projekt.program.Organisms.Plants.Dandelion;
import projekt.program.Organisms.Plants.DeadlyNightshade;
import projekt.program.Organisms.Plants.Guarana;

import javax.swing.*;
import java.awt.*;

public class Panel extends JPanel {
    private int RECT_SIZE;
    int width, height;
    World swiat;
    public Panel(int height, int width, int RECT_SIZE, World swiat){
        setPreferredSize(new Dimension(width*RECT_SIZE+1, (10+height)*RECT_SIZE+10));
        this.width = width;
        this.height = height;
        this.swiat = swiat;
        this.RECT_SIZE = RECT_SIZE;
    }
    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2d = (Graphics2D) g;

        for(int i = 0; i < height; i++){
            for(int j = 0; j < width; j++){
                //g2d.drawRect(j*RECT_SIZE, i*RECT_SIZE, RECT_SIZE, RECT_SIZE);
                // Get the organism at the current position
                Organism organism = swiat.getOrganismAtPosition(j, i);
                if(organism instanceof Animal){
                    g2d.setColor(getColorForOrganism(organism));
                    g2d.fillOval(j*RECT_SIZE, i*RECT_SIZE, RECT_SIZE, RECT_SIZE);
                    g2d.setColor(Color.BLACK);
                    g2d.drawOval(j*RECT_SIZE, i*RECT_SIZE, RECT_SIZE, RECT_SIZE);
                }
                if(organism instanceof Plant) {
                    g2d.setColor(Color.GREEN);
                    g2d.fillRect(j * RECT_SIZE, i * RECT_SIZE, RECT_SIZE, RECT_SIZE);
                    g2d.setColor(getColorForOrganism(organism));
                    for (int k = 0; k < 4; k++) {
                        int dotX = j * RECT_SIZE + RECT_SIZE / 4 + (k % 2) * RECT_SIZE / 2 - RECT_SIZE / 8;
                        int dotY = i * RECT_SIZE + RECT_SIZE / 4 + (k / 2) * RECT_SIZE / 2 - RECT_SIZE / 8;
                        if(organism instanceof DeadlyNightshade || organism instanceof Guarana || organism instanceof Barszcz){
                            g2d.fillOval(dotX, dotY, RECT_SIZE / 6, RECT_SIZE / 6);
                        }
                        if(organism instanceof Dandelion){
                            g2d.drawLine(dotX,dotY,dotX+ RECT_SIZE / 6, dotY+RECT_SIZE / 6);
                            g2d.drawLine(dotX+ RECT_SIZE / 6,dotY,dotX, dotY+RECT_SIZE / 6);
                            g2d.drawLine(dotX+ RECT_SIZE / 12, dotY, dotX+RECT_SIZE / 12, dotY+RECT_SIZE / 6);
                            g2d.drawLine(dotX, dotY+ RECT_SIZE / 12, dotX+RECT_SIZE / 6, dotY+RECT_SIZE / 12);
                        }
                    }
                    g2d.setColor(Color.BLACK);
                }
            }
        }
        g2d.drawRect(0, 0, RECT_SIZE*width, RECT_SIZE*height);
        g2d.drawRect(0,RECT_SIZE*(width+1), RECT_SIZE*width, RECT_SIZE*3);
        String message = "Aktualna tura: " +swiat.getTurn();
        g2d.drawString(message, RECT_SIZE,RECT_SIZE*(width+2));
        //g2d.drawString(swiat.getMessage(), RECT_SIZE,RECT_SIZE*(width+2)+10);
    }
    // Method to get the color for each type of organism
    private Color getColorForOrganism(Organism organism) {
        //animals
        if (organism instanceof Wolf) {
            return Color.GRAY;
        } else if (organism instanceof Sheep) {
            return Color.WHITE;
        } else if (organism instanceof Turtle) {
            return Color.green.darker().darker();
        } else if (organism instanceof Fox) {
            return Color.ORANGE;
        } else if (organism instanceof Antelope) {
            return Color.YELLOW;
        } else if (organism instanceof Human) {
            return Color.PINK;
        }
        //plants
        else if (organism instanceof Guarana) {
            return Color.red;
        } else if (organism instanceof Dandelion) {
            return Color.yellow;
        } else if (organism instanceof DeadlyNightshade) {
            return Color.MAGENTA;
        } else if (organism instanceof Barszcz) {
            return Color.white;
        }

        // Default color if the organism type is unknown
        return Color.BLACK;
    }
}
